protocol/outline/outbound.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/sagernet/sing-box/adapter/outbound"
+	"github.com/sagernet/sing-box/common/dialer"
+	C "github.com/sagernet/sing-box/constant"
+	"github.com/sagernet/sing-box/log"
+	"github.com/sagernet/sing-box/option"
+	"github.com/sagernet/sing/common/logger"
+	"github.com/sagernet/sing/common/metadata"
+	"github.com/sagernet/sing/common/network"
+	"gopkg.in/yaml.v3"
+)
+
+// Outbound implements the smart dialer outbound from outline sdk
+type Outbound struct {
+	outbound.Adapter
+	logger       logger.ContextLogger
+	dialer       transport.StreamDialer
+	dialerMutex  *sync.Mutex
+	createDialer func() (transport.StreamDialer, error)
+}
+
+// RegisterOutbound registers the outline outbound to the registry
+func RegisterOutbound(registry *outbound.Registry) {
+	outbound.Register[option.OutboundOutlineOptions](registry, C.TypeOutline, NewOutbound)
+}
+
+// NewOutbound creates a proxyless outbond that uses the proxyless transport
+// for dialing
+func NewOutbound(ctx context.Context, router adapter.Router, log log.ContextLogger, tag string, options option.OutboundOutlineOptions) (adapter.Outbound, error) {
+	outboundDialer, err := dialer.New(ctx, options.DialerOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if options.TestTimeout == "" {
+		options.TestTimeout = "0s"
+	}
+	timeout, err := time.ParseDuration(options.TestTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	outboundStreamDialer := &outboundStreamDialer{
+		dialer: outboundDialer,
+		logger: log,
+	}
+
+	strategyFinder := &smart.StrategyFinder{
+		TestTimeout:  timeout,
+		LogWriter:    &logWriter{log},
+		StreamDialer: outboundStreamDialer,
+		PacketDialer: outboundStreamDialer,
+	}
+
+	yamlOptions, err := yaml.Marshal(options)
+	if err != nil {
+		return nil, err
+	}
+	outbound := &Outbound{
+		Adapter:     outbound.NewAdapterWithDialerOptions(C.TypeOutline, tag, []string{network.NetworkTCP}, options.DialerOptions),
+		logger:      log,
+		dialerMutex: &sync.Mutex{},
+		// During the dialer creation the strategy finder try to use the stream dialer
+		// for resolving the domains. We can't create the smart dialer during the
+		// outbound initialization because there wouldn't be a tunnel to communicate.
+		// So for fixing this issue, the dialer must be created during the DialContext call.
+		createDialer: sync.OnceValues(func() (transport.StreamDialer, error) {
+			dialer, err := strategyFinder.NewDialer(ctx, options.Domains, yamlOptions)
+			if err != nil {
+				return nil, err
+			}
+			return dialer, nil
+		}),
+	}
+
+	return outbound, nil
+}
+
+// DialContext extracts the metadata domain, add the destination to the context
+// and use the proxyless dialer for sending the request
+func (o *Outbound) DialContext(ctx context.Context, network string, destination metadata.Socksaddr) (net.Conn, error) {
+	o.dialerMutex.Lock()
+	if o.dialer == nil {
+		dialer, err := o.createDialer()
+		if err != nil {
+			return nil, err
+		}
+		o.dialer = dialer
+	}
+	o.dialerMutex.Unlock()
```

Comment:
I think you need to defer the Unlock above, as otherwise the early return will never release it

Reply:
I can't defer since that would lock the dialer every time we got a request, but I'll unlock if received an error and that will unlock in all cases